File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> freq = new HashMap<>();
        for(int n : nums){
            freq.put(n, freq.getOrDefault(n, 0) + 1);
        }
        List<Integer>[] buckets = new List[nums.length+1];
        for(int i = 0; i < nums.length+1; i++){
            buckets[i] = new ArrayList();
        }

        for(Map.Entry<Integer, Integer> entry : freq.entrySet()){
            buckets[entry.getValue()].add(entry.getKey());
        }

        int[] res = new int[k];
        int idx = 0;
        for(int i = buckets.length-1; i >= 0; i--){
            for(int n : buckets[i]){
                System.out.println("N: " + n);
                res[idx++] = n;
                if(idx >= k){
                    return res;
                }
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/4sum/submission-2.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        nums.sort()

        i = 0

        while i < len(nums) - 3:
            x = i + 1
            while x < len(nums) - 2:
                l = x + 1
                r = len(nums) - 1
                while l < r:
                    val = nums[i] + nums[x] + nums[l] + nums[r]
                    if val > target:
                        r -= 1
                    elif val < target:
                        l += 1
                    else:
                        res.append([nums[i], nums[x], nums[l], nums[r]])
                        l += 1
                        r -= 1
                        while l < len(nums) and nums[l] == nums[l-1]:
                            l +=1
                        while r > l and nums[r] == nums[r+1]:
                            r -= 1
                x += 1
                while x < len(nums) - 2 and nums[x] == nums[x-1]:
                    x += 1
            i += 1
            while i < len(nums) -3 and nums[i] == nums[i-1]:
                i +=1        

        return res